<img src='https://i.imgur.com/RDAD11M.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 2 / NOTE 1*

# 📝 Assignment

# 가설 검정

## 1. One-Sample T-test 연습

### 데이터 불러오기
- [서울시 가로수 데이터셋](https://data.seoul.go.kr/dataList/367/S/2/datasetView.do)을 사용합니다.

다음 링크를 참조하세요. 
<https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/seoul_tree/seoul_tree.txt>


In [107]:
# 이 곳에 답안을 작성하시기 바랍니다.

import pandas as pd

trees = pd.read_csv('https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/seoul_tree/seoul_tree.txt',  sep = '\t', skiprows = 1)
trees = trees.replace({'-':0})

print(trees.head())
print(trees.shape) # (28, 24)

     기간  자치구    노선수       수량     은행나무  ...    무궁화    소나무   목련    이팝나무      기타
0  2019  서울시  1,412  307,351  108,050  ...  2,374  4,539  378  20,088  20,112
1  2019  종로구     70    7,329    4,072  ...      0    239    0     711      29
2  2019   중구     48    8,250    3,640  ...     46  2,162    0     470     817
3  2019  용산구     39    7,835    4,444  ...      0     51  126     198     780
4  2019  성동구     38    8,120    1,090  ...      0     53    0     709     646

[5 rows x 24 columns]
(28, 24)


### "서울시에는 구별로 평균 약 400 그루 정도의 `이팝나무`가 있다"라는 가설에 대해 가설검정을 시행하세요.

- 단, `서울시`, `공원녹지사업소`, `시설관리공단` 부분의 데이터는 제외합니다.
- 가설검정 결과의 pvalue 값을 `pv1`에 저장하세요. (`.pvalue` 사용)

In [108]:
# 이 곳에 답안을 작성하시기 바랍니다

trees['이팝나무'] = pd.to_numeric(trees['이팝나무'].str.replace(',','')) 
trees['이팝나무'].head()

0    20088
1      711
2      470
3      198
4      709
Name: 이팝나무, dtype: int64

In [109]:
trees[ (trees['자치구'] == '서울시') | (trees['자치구'] == '공원녹지사업소') | (trees['자치구'] == '시설관리공단') ].index 

Int64Index([0, 26, 27], dtype='int64')

In [110]:
idx = trees[ (trees['자치구'] == '서울시') | (trees['자치구'] == '공원녹지사업소') | (trees['자치구'] == '시설관리공단') ].index 
trees = trees.drop(idx)
trees.shape

(25, 24)

In [111]:
trees['이팝나무'].describe()

count      25.00000
mean      688.64000
std       426.86599
min        47.00000
25%       443.00000
50%       642.00000
75%       853.00000
max      1651.00000
Name: 이팝나무, dtype: float64

In [112]:
# 가설 검정 #*
from scipy import stats
result = stats.ttest_1samp(trees['이팝나무'], 400)   # 데이터, 비교하려는 값

print('t검정 통계량 = %.3f, pvalue = %.3f'%(result)) 

t검정 통계량 = 3.381, pvalue = 0.002


* 귀무가설 : m = 400 ( 이팝나무의 서울시 구별 평균 나무 수는 400그루이다. )
* 대립가설 : m =! 400 ( 이팝나무의 서울시 구별 평균 나무 수가 400그루가 아니다. )

pvalue = 0.002 < 0.05 이므로 유의수준 5%에서 귀무가설 기각, 이는 서울시 구별 이팝나무의 평균 나무 수는 400그루가 아니라는 강한 증거가 된다. )


In [133]:
# t값
( 688.64 - 400 ) / ( 426.866 / 5 )
# 자유도가 24인 t분포에서 검정통계량인 3.38은 0.002의 유의확률을 가진다.

3.3809204762150182

In [113]:
pv1 = result.pvalue
pv1

0.002470712850935819

In [114]:
# trees.iloc[24,:]

### 이번에는 10개의 구를 샘플링하여, 위에 대한 가설검정을 다시 한번 시행해보세요.

- 단, 아래의 코드를 참조하여 sample을 위한 seed 는 123로 설정하세요.
- 가설검정 결과의 pvalue값을 `pv2`에 저장하세요


In [115]:
import numpy as np
np.random.seed(123)
 
idx = np.random.choice(24, size = 10, replace=False)  # "~"와 " "을 적절한 값으로 채우시면 됩니다
idx

array([ 5, 20, 21, 18, 15,  8,  7, 11,  4,  3])

In [116]:
trees_10 = trees.iloc[idx,:]

result = stats.ttest_1samp(trees_10['이팝나무'], 400)   # 데이터, 비교하려는 값

print('t검정 통계량 = %.3f, pvalue = %.3f'%(result)) # < 0.05 이므로 귀무가설 기각, 이팝나무의 서울시 자치구별 평균 나무 수가 400그루라는 가설은 유의수준 5%에서  유의하지 않다. 

t검정 통계량 = 3.160, pvalue = 0.012


* 귀무가설 : m = 400 ( 이팝나무의 서울시 구별 평균 나무 수는 400그루이다. )
* 대립가설 : m =! 400 ( 이팝나무의 서울시 구별 평균 나무 수가 400그루가 아니다. )

pvalue = 0.012 < 0.05 이므로 유의수준 5%에서 귀무가설 기각, 이는 서울시 구별 이팝나무의 평균 나무 수는 400그루가 아니라는 강한 증거가 된다. )

In [134]:
trees_10.describe()

기간         이팝나무
count    10.0    10.000000
mean   2019.0   791.800000
std       0.0   392.091486
min    2019.0   249.000000
25%    2019.0   478.250000
50%    2019.0   734.500000
75%    2019.0  1095.250000
max    2019.0  1426.000000

In [136]:
import numpy as np

(791.8 - 400) / (392.0915 / np.sqrt(10)) # t값이 3.16과 비슷함을 확인. 자유도가 9인 t분포에서 3.16의 검정통계량은 0.012의 피벨류를 가진다. < 0.05 -> 귀무가설 기각. 

3.1599266682750606

In [117]:
pv2 = result.pvalue
pv2

0.01155157903647694

## 2. Two-Sample T-test 연습

### "서울시의 구별 평균 느티나무수와 왕벚나무의 수는 차이가 없다"라는 가설에 대해 가설검정을 시행하세요.

- 단, `서울시`, `공원녹지사업소`, `시설관리공단` 부분의 데이터는 제외합니다.
- 가설검정 결과의 pvalue를 `pv3`에 저장하세요

In [118]:
trees['느티나무'] = pd.to_numeric(trees['느티나무'].str.replace(',','')) 
trees['왕벚나무'] = pd.to_numeric(trees['왕벚나무'].str.replace(',','')) 


In [137]:
# t = pd.Dataframe( trees['느티나무'], trees['왕벚나무'] )
trees.loc[:, ( ['느티나무', '왕벚나무'] ) ].describe()



느티나무         왕벚나무
count    25.000000    25.000000
mean   1415.120000  1292.000000
std    1036.783862   832.740106
min     185.000000   254.000000
25%     619.000000   602.000000
50%    1057.000000  1096.000000
75%    2231.000000  2095.000000
max    3419.000000  2952.000000

In [139]:
# 복수표본 t분포 
(1415.12-1292 ) / np.sqrt(    (1036.783862**2 / 25) + (832.740106**2 / 25) )
# t값이 같음을 확인.    

0.46292560136489425

In [120]:
# 이 곳에 답안을 작성하시기 바랍니다

result = stats.ttest_ind(trees['느티나무'], trees['왕벚나무'])
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result))
# 귀무가설 채택. 즉, 유의수준 5%에서 서울시 자치구별 느티나무와 왕벚나무의 평균 수는 차이가 유의마하게 없다.

t검정 통계량 = 0.463, pvalue = 0.646


* 귀무가설 : m = 0 ( 차이가 없다. = 느티나무와 왕벚나무의 평균에 차이가 없다.)

* 대립가설 : m =! 0  (차이가 있다. = 평균 느티나무 수가 왕벚나무 수의 평균은 같지 않다.)

'같지 않다'라는 내용의 대립 가설 이므로 두 평균의 차이가 음수이든 양수이든 상관이 없다. 즉,  양측 가설 검정을 한다.  

 pvalue = 0.646 > 0.05 이므로 유의수준 5%에서 귀무가설을 기각할 수 없다. 즉, 귀무가설 채택한다. 대립가설을 강하게 뒷받침하는 데이터가 아니므로 반대의 경우에도 귀무가설을 기각하지 못한다. #( 즉, 두 나무의 평균 그루 수에 차이가 별로 없다. ) #

In [121]:
pv3 = result.pvalue
pv3

0.6455096880085703


### Two-Sample One-tailed test를 사용하여 **평균 느티나무수가 왕벚나무 수보다 크다**라는 가설에 대한 가설검정을 시행하세요.
- 해당 가설검정의 결과를 `ans`에 저장하세요.

이 때 `ans`가 가질 수 있는 값은 아래와 같습니다 :
1. "느티나무" ( 느티나무의 수가 통계적으로 유의하게 많은 경우 )
2. "왕벚나무" ( 왕벚나무의 수가 통계적으로 유의하게 많은 경우 )
3. "없음" (통계적으로 유의하지 않은 경우)

다음 링크를 참조하세요.
<https://stackoverflow.com/questions/15984221/how-to-perform-two-sample-one-tailed-t-test-with-numpy-scipy>

귀무가설 : m = 0 ( 차이가 없다. = 느티나무와 왕벚나무의 평균에 차이가 없다.)

대립가설 : m > 0  (차이가 있다. = 평균 느티나무 수가 왕벚나무 수보다 크다.)




In [122]:
!pip install --upgrade scipy
import scipy
scipy.version

Requirement already up-to-date: scipy in /usr/local/lib/python3.7/dist-packages (1.6.3)


<module 'scipy.version' from '/usr/local/lib/python3.7/dist-packages/scipy/version.py'>

In [123]:
# 이 곳에 답안을 작성하시기 바랍니다


result = stats.ttest_ind(trees['느티나무'], trees['왕벚나무'])

result = result.pvalue / 2
print(' pvalue = %.3f'%(result))

 pvalue = 0.323


* 귀무가설 : m = 0 ( 차이가 없다. = 느티나무와 왕벚나무의 평균에 차이가 없다.)

* 대립가설 : m > 0  (차이가 있다. = 평균 느티나무 수가 왕벚나무 수보다 크다.)

단측 가설 검정 결과 

 pvalue = 0.323 > 0.05 이므로 유의수준 5%에서 귀무가설을 기각할 수 없다. 즉, 귀무가설 채택한다. 대립가설을 강하게 뒷받침하는 데이터가 아니므로 반대의 경우에도 귀무가설을 기각하지 못한다. #( 즉, 두 나무의 평균 그루 수에 차이가 별로 없다. ) #

In [124]:
ans = "없다" 
ans

'없다'

In [125]:

result = stats.ttest_ind(trees['왕벚나무'], trees['느티나무'])
print('t검정 통계량 = %.3f, pvalue = %.3f'%(result))

t검정 통계량 = -0.463, pvalue = 0.646


In [132]:
pv3


0.6455096880085703

## 🔥 **도전과제**
> 3문제 모두 하면 됩니다.



### 1. Compare mean for multiple-sample 
 
다음 링크를 참조하세요. <https://reneshbedre.github.io/blog/anova.html>

구별, 은행나무, 양버즘나무, 느티나무에 대해서 ANOVA Test를 실행하세요.

단, `서울시`, `공원녹지사업소`, `시설관리공단` 부분의 데이터는 제외합니다.

In [126]:
# 이 곳에 답안을 작성하시기 바랍니다



### 2. Sampling

- 세션 시간에 배웠던 4가지 sample 방식을 구현해보세요.

- 이를 위한 데이터는 1 부터 100 이며, sampling해야 하는 수는 20개 입니다.

- Stratified random sampling을 위한 그룹을 나누는 기준은 1-10, 11-20, ... 과 같은 방식입니다.

- Cluster sampling 을 위한 그룹을 나누는 기준은 5로 나눴을때의 나머지 값입니다. (0,1,2,3,4)

In [127]:
# 이 곳에 답안을 작성하시기 바랍니다

### 3. Data transform
다음 링크를 참조하여  `Inverse transform sampling` 을 구현합니다.

<https://youtu.be/9ixzzPQWuAY>

0 ~ 1 의 값을 동일한 확률로 갖는 `uniform distribution`으로 부터 `binomial distribution`으로 transform 하는 코드를 작성한 뒤, 두가지 결과를 시각화를 통해 비교하세요.


In [128]:
# 이 곳에 답안을 작성하시기 바랍니다